src/avram/bulk_upsert.cr

```diff
@@ -0,0 +1,93 @@
+class Avram::BulkUpsert(T)
+  def initialize(@records : Array(T),
+                 @conflicts : Array(Symbol),
+                 @permitted_fields : Array(Symbol))
+    @records = set_timestamps(records)
+    @sample_record = @records.first.as(T)
+    @permitted_fields = permitted_fields_for(permitted_fields)
+  end
+
+  def statement
+    <<-SQL
+    INSERT INTO #{table}(#{fields})
+      (SELECT * FROM unnest(#{value_placeholders}))
+      ON CONFLICT (#{conflicts}) DO UPDATE SET #{updates}
+      RETURNING #{returning}
+    SQL
+  end
+
+  def args
+    @records.map do |record|
+      permitted_attributes(record).map(&.value)
+    end.transpose
+  end
+
+  private def permitted_fields_for(fields : Array(Symbol))
+    fields.push(:created_at) if @sample_record.responds_to?(:created_at)
+    fields.push(:updated_at) if @sample_record.responds_to?(:updated_at)
+    fields.uniq
+  end
+
+  private def permitted_attributes(record)
+    record
+      .attributes
+      .select { |attr| @permitted_fields.includes?(attr.name) }
+  end
+
+  private def permitted_attributes
+    permitted_attributes(@sample_record)
+  end
+
+  private def conflicts
+    @conflicts.join(", ")
+  end
+
+  private def set_timestamps(collection)
+    collection.map do |record|
```

Comment:
This could be a `.each` and won't need to have the last line of the block returning the record